torch/nn/modules/module.py

```diff
@@ -192,10 +192,11 @@ def _apply(self, fn):
         for param in self._parameters.values():
             if param is not None:
                 # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                # want to create copy nodes, so we have to use `torch.no_grad`.
+                with torch.no_grad():
+                    param = fn(param)
+                    if param._grad is not None:
+                        param._grad = fn(param._grad)
```

Comment:
I need to debug why we need this change

Reply:
Update: For some test cases, `param._grad` is already a tensor with `allow_tensor_metadata_change` set to false, and changing its storage in `set_data()` will throw an error. To fix this problem, we should just do `param._grad = fn(param._grad)`, and use `with torch.no_grad()` to avoid accumulating gradients.

Reply:
New code is more idiomatic anyway, sgtm.